status-im/status-react
# Redirect user to Activity Center to manage pending contact requests

fixes #14591
fixes #14579 

### Summary

- [x] `Fixed` https://github.com/status-im/status-mobile/issues/14591
- [x] https://github.com/status-im/status-mobile/issues/14579
- [x] Removed code for the old bottom sheet screen to manage pending contact requests.
- [x] Replaced usages of `status-im.utils.fx` with `utils.re-frame/dispatch` (spoiler, this was much harder than it seems, see the notes below if you're curious).
- [x] Refactored test helper for testing logs in order to better work with asynchronous ClojureScript tests.

#### New behavior

>Given a user has at least 1 pending contact request,
>And they open the `Messages` screen > `Contacts` tab,
>And they press the `Pending requests` section,
>Then the Activity Center should be opened in the `Contact requests` tab with the `unread` filter enabled.

[Pending contact requests.webm](https://user-images.githubusercontent.com/46027/208979440-05a72396-04e1-4711-9cee-d3ea9c3821ab.webm)

### Notes about testing event handlers using our `utils.re-frame/dispatch` function

In order to write tests for event handlers using `re-frame.test/run-test-sync` we need to make sure all calls to `re-frame.core/dispatch` are visible to the macro. The thing is that our guidelines suggest we use `utils.re-frame/dispatch` instead, but this doesn't work with the testing library.

The current implementation in https://github.com/day8/re-frame-test/blob/master/src/day8/re_frame/test.cljc#L297 uses `with-redefs`, which **does not work with our aliased function `utils.re-frame/dispatch`**. This is how `with-redefs` work, there's nothing we can do.

This PR solves this problem by wrapping the original `re-frame.test/run-test-sync` macro body with our own `with-redefs` to explicitly redefine our aliased `utils.re-frame/dispatch` function.

Hence, from now on, we're able to write tests for event handlers that use our `utils.re-frame/dispatch` function, so no more using `re-frame.core/dispatch`. cc @flexsurfer

### Notes about the improved way to test logs

During RDD (REPL-Driven Development) sessions, I have faced limitations to write tests for log messages when the tests are asynchronous, because it was harder to inspect logs when assertions failed. I solved this problem by creating a new fixture.

Therefore, for anyone that wants to test log messages (in event handler namespaces or not), as this PR shows, you just need to call `(status-im.test-helpers/use-log-fixture)` at the top of the namespace. I think the usage is now even simpler and more flexible!

### Platforms

- Android
- iOS

### Steps to test

1. Create a new account A or use an account without contact requests (pending or not).
2. Send a contact request from B to A.
3. Verify the pending request is displayed in `Messages` screen > `Contacts` tab.
4. Click on the pending request and verify the Activity Center is opened in the `Contact requests` tab with the `unread` filter enabled.
5. Switch tabs in the Activity Center and close it.
6. Click again on the `Pending requests` section and notice the Activity Center always opens with the `Contact requests` tab opened. This allows the user to easily manage contact requests.

status: ready
